src/week2.hs

```diff
@@ -0,0 +1,176 @@
+         deriving (Show, Eq, Ord)
+
+-- A code is defined to simply be a list of Pegs
+type Code = [Peg]
+
+-- A move is constructed using a Code and two integers; the number of
+-- exact matches and the number of regular matches
+data Move = Move Code Int Int
+          deriving (Show, Eq)
+
+-- List containing all of the different Pegs
+colors :: [Peg]
+colors = [Red, Green, Blue, Yellow, Orange, Purple]
+
+-- Exercise 1 -----------------------------------------
+
+-- Get the number of exact matches between the actual code and the guess
+exactMatches :: Code -> Code -> Int
+
+exactMatches [] _ = 0
+exactMatches _ [] = 0
+exactMatches (x:xs) (x':xs') 
+    |x == x' = 1 + (exactMatches xs xs')
+    |otherwise = (exactMatches xs xs')
+
+-- Exercise 2 -----------------------------------------
+-- Write a function that returns the number of total matches between the secret code and the guess
+
+-- helper
+sumLists :: [Int] -> [Int] -> [Int]
+sumLists [] _ = []
+sumLists _ [] = []
+sumLists (x:xs) (x':xs') = (x + x'):(sumLists xs xs')
+
+--or, with implicit recursion (Jan's suggestion):
+sumLists1 :: [Int] -> [Int] -> [Int]
+sumLists1 a b = zipWith (+) a b
+
+countColors :: Code -> [Int]
+
+countColors [] = [0,0,0,0,0,0]
+countColors (x:xs)
+    | x == Red    = sumLists [1,0,0,0,0,0] (countColors xs)
+    | x == Green  = sumLists [0,1,0,0,0,0] (countColors xs)
+    | x == Blue   = sumLists [0,0,1,0,0,0] (countColors xs)
+    | x == Yellow = sumLists [0,0,0,1,0,0] (countColors xs)
+    | x == Orange = sumLists [0,0,0,0,1,0] (countColors xs)
+    | otherwise   = sumLists [0,0,0,0,0,1] (countColors xs)
+
+sumMin :: [Int] -> [Int] -> Int
+sumMin [] _ = 0
+sumMin _ [] = 0
+sumMin (x:xs) (x':xs') = (min x x') + (sumMin xs xs')
+
+-- or, with implicit recursion and more concisely (Jan's suggestion):
+sumMin1 :: [Int] -> [Int] -> Int
+sumMin1 a b = sum (zipWith (min) a b)
+
+matches :: Code -> Code -> Int
+matches x y = sumMin (countColors x) (countColors y) 
+
+-- Exercise 3 -----------------------------------------
+-- Construct a Move from a guess given the actual code
+getMove :: Code -> Code -> Move
+getMove s g = Move (g) (exactMatches s g)  ((matches s g)  - (exactMatches s g))
+
+-- Or:
+getMove1 :: Code -> Code -> Move
+getMove1 s g = Move g e (m-e) 
+        where e = exactMatches s g
+              m = matches s g
+
+-- Exercise 4 -----------------------------------------
+-- A code is consistent with a Move if the Code could have been the secret that generated that move
+-- Write the function isConsistent
+isConsistent :: Move -> Code -> Bool
+
+isConsistent (Move c x y) c' 
+    |(exactMatches c c') == x && ((matches c c') - (exactMatches c c')) == y = True
+    |otherwise                                                               = False
+    
+-- Or:
+isConsistent1 :: Move -> Code -> Bool
+isConsistent1 (Move c x y) c'  = e == x && (m - e) == y
+    where e = exactMatches c c'
+          m = matches c c'
```

Comment:
this is definitely the preferred choice---if you're doing control flow on a Boolean expression, it's best to just return that Boolean expression itself